File: NET/JNetCall.Sharp.Common/Impl/IO/Net/TcpTransport.cs
using System;
using System.Net;
using System.Net.Sockets;
using System.Threading;
using JNetCall.Sharp.API.Enc;
using JNetCall.Sharp.API.IO;

namespace JNetCall.Sharp.Impl.IO.Net
{
    public sealed class TcpTransport : ISendTransport, IPullTransport, IDisposable
    {
        private readonly IEncoding<byte[]> _encoding;
        private readonly TcpListener _receiver;
        private readonly TcpClient _sender;
        private readonly IPEndPoint _endPointOut;
        private readonly int _wait;

        public TcpTransport(IEncoding<byte[]> encoding,
            string hostIn, int portIn, string hostOut, int portOut)
            : this(encoding, NetworkTools.ToEndPoint(hostIn, portIn),
                NetworkTools.ToEndPoint(hostOut, portOut))
        {
        }

        private TcpTransport(IEncoding<byte[]> encoding,
            IPEndPoint endPointIn, IPEndPoint endPointOut)
        {
            _encoding = encoding;
            _receiver = new TcpListener(endPointIn);
            _receiver.Start();
            _sender = new TcpClient();
            _endPointOut = endPointOut;
            _wait = 50;
        }

        private void ForceConnect()
        {
            while (!_sender.Connected)
            {
                _sender.Connect(_endPointOut);
                if (_sender.Connected)
                    break;
                Thread.Sleep(_wait);
            }
        }

        private void ForceAccept()
        {
            while (_receiverConn == null)
            {
                _receiverConn = _receiver.AcceptTcpClient();
                if (_receiverConn != null)
                    break;
                Thread.Sleep(_wait);
            }
        }

        private readonly object _sendSync = new();

        public void Send<T>(T payload)
        {
            lock (_sendSync)
            {
                ForceConnect();
                var bytes = _encoding.Encode(payload);
                _sender.GetStream().Write(bytes, 0, bytes.Length);
            }
        }

        private readonly object _receiveSync = new();
        private TcpClient _receiverConn;

        public T Pull<T>()
        {
            lock (_receiveSync)
            {
                ForceAccept();
                var bytes = StreamTools.ReadWithSize(_receiverConn.GetStream());
                return _encoding.Decode<T>(bytes);
            }
        }

        public void Dispose()
        {
            _encoding.Dispose();
            _sender.Close();
            _sender.Dispose();
            _receiverConn?.Close();
            _receiverConn?.Dispose();
            _receiver.Stop();
        }
    }
}

File: JVM/jnetcall-java-common/src/main/java/jnetcall/java/impl/io/net/TcpTransport.java
package jnetcall.java.impl.io.net;

import jnetcall.java.api.enc.IEncoding;
import jnetcall.java.api.io.IPullTransport;
import jnetcall.java.api.io.ISendTransport;

import java.io.IOException;
import java.net.SocketAddress;
import java.nio.ByteBuffer;
import java.nio.channels.ServerSocketChannel;
import java.nio.channels.SocketChannel;

public final class TcpTransport implements ISendTransport, IPullTransport, AutoCloseable {

    private final IEncoding<byte[]> _encoding;
    private final ServerSocketChannel _receiver;
    private final SocketChannel _sender;
    private final SocketAddress _endPointOut;
    private final int _wait;

    public TcpTransport(IEncoding<byte[]> encoding,
                         String hostIn, int portIn, String hostOut, int portOut) {
        this(encoding, NetworkTools.toEndPoint(hostIn, portIn),
                NetworkTools.toEndPoint(hostOut, portOut));
    }

    private TcpTransport(IEncoding<byte[]> encoding,
                          SocketAddress endPointIn, SocketAddress endPointOut) {
        _encoding = encoding;
        try {
            _receiver = ServerSocketChannel.open().bind(endPointIn);
            _sender = SocketChannel.open();
            _endPointOut = endPointOut;
            _wait = 50;
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    private void forceConnect() throws IOException, InterruptedException {
        while (!_sender.isConnected()) {
            _sender.connect(_endPointOut);
            if (_sender.isConnected())
                break;
            Thread.sleep(_wait);
        }
    }

    private void forceAccept() throws IOException, InterruptedException {
        while (_receiverConn == null) {
            _receiverConn = _receiver.accept();
            if (_receiverConn != null)
                break;
            Thread.sleep(_wait);
        }
    }

    private final Object _sendSync = new Object();

    @Override
    public <T> void send(T payload) {
        try {
            synchronized (_sendSync) {
                forceConnect();
                var bytes = _encoding.encode(payload);
                var buff = ByteBuffer.wrap(bytes);
                _sender.write(buff);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private final Object _receiveSync = new Object();
    private SocketChannel _receiverConn;

    @Override
    public <T> T pull(Class<T> clazz) {
        try {
            synchronized (_receiveSync) {
                forceAccept();
                var buff = NetworkTools.readWithSize(_receiverConn);
                var bytes = buff.array();
                return _encoding.decode(bytes, clazz);
            }
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    public void close() throws Exception {
        _encoding.close();
        _sender.close();
        if (_receiverConn != null) _receiverConn.close();
        _receiver.close();
    }
}
